Userland/Libraries/LibThreading/ConditionVariable.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2021, kleines Filmröllchen <malu.bertsch@gmail.com>.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <LibThreading/Mutex.h>
+#include <pthread.h>
+#include <sys/cdefs.h>
+#include <sys/types.h>
+
+namespace Threading {
+
+// A signaling condition variable that wraps over the pthread_cond_* APIs.
+class ConditionVariable {
+    friend class Mutex;
+
+public:
+    ConditionVariable()
+    {
+        pthread_cond_init(&m_condition, nullptr);
+    }
+
+    ALWAYS_INLINE ~ConditionVariable()
+    {
+        pthread_cond_destroy(&m_condition);
+    }
+
+    // As with pthread APIs, to_wait_on must be locked or undefined behavior ensues.
+    ALWAYS_INLINE void wait_on(Mutex& to_wait_on)
+    {
+        pthread_cond_wait(&m_condition, &to_wait_on.m_mutex);
+    }
```

Comment:
What would be really nice here is a method accepting a lambda to check the condition. Here's how it looks [in LWSP](https://github.com/bugaevc/lets-write-sync-primitives/blob/833be0d5e932c29a34d3dae8e01dccbbb0393570/src/condvar.cpp#L26):

```cpp
void CondVar::wait(std::function<bool()> condition) {
    while (!condition()) {
        wait();
    }
}
```

It's a small thing, but it makes actually using the condition variable so much cleaner and less error-prone.

Reply:
On that note, should `Threading::ConditionVariable` store a reference to its mutex, like LWSP condvar does, instead of requiring a mutex argument in `wait()` calls? I think that would be nicer.

Reply:
Yeah, I think I'll make it "connect" to a mutex "permanently".

I'm not sure what to take from your lambda suggestion. The wait() call doesn't return until it gets signalled, so why would you want a loop around it?

Reply:
No, `wait()` can return spuriously, that's just how it is. To quote POSIX:

> Spurious wakeups from the `pthread_cond_wait()` or `pthread_cond_timedwait()` functions may occur. Since the return from `pthread_cond_wait()` or `pthread_cond_timedwait()` does not imply anything about the value of this predicate, the predicate should be re-evaluated upon such return.

Moreover, even if `wait()` returns because of someone actually signaling the condition variable (and not spuriously), that doesn't yet mean the condition is now true: others may have changed it again before we grabbed the mutex.

So no, you always have to re-check the condition upon returning from `wait()`. And this is what I mean when I say wait-with-lambda makes this less error prone :slightly_smiling_face: